File: abstractFactory/python/Store.py
from OSXStore import OSXStore
from interfaces import StoreInstance
from WindowsStore import WindowsStore

class Store(StoreInstance):
  def __init__(self, platform):
    self.__apps = []  
    self.__games = []

    if platform == 'windows':
      self.__factory = WindowsStore()
    if platform == 'osx':
      self.__factory = OSXStore()
    pass

  def __getInstance(self, type, name):
    instances = None
    if type == 'app':
      instances = self.__apps
    if type == 'game':
      instances = self.__games
    
    for instance in instances:
      if instance.getInfo()['name'] == name:
        return instance
    
    return None

  def run(self, type, name):
    targetInstance = self.__getInstance(type, name)
    if targetInstance == None:
      raise ValueError('instance is not define')
    
    targetType = targetInstance.getInfo()['type']
    
    if targetType == 'app':
      targetInstance.start()
    if targetType == 'game':
      targetInstance.play()

    pass

  def joinGame(self, name):
    newGame = self.__factory.createGame(name)
    self.__games.append(newGame)
    pass
  
  def buyApp(self, name, deposit):
    newApp = self.__factory.createApplication(name)
    if newApp.buyApp(deposit):
      self.__apps.append(newApp)
    
    pass

  def printInfo(self, type, name):
    targetInstance = self.__getInstance(type, name)
    if targetInstance == None:
      raise ValueError('instance is not define')

    targetInstance.printInformation()
    pass

File: abstractFactory/typescript/Store.ts
import {
  AppInstance,
  StoreFactory,
  GameInstance,
  StoreInstance
} from './interfaces'

import OSXStore from './OSXStore'
import WindowsStore from './WindowsStore'

export default class Store implements StoreInstance {
  private apps: AppInstance[]
  private games: GameInstance[]
  private factory: StoreFactory
  
  constructor(platform: string) {
    this.apps = [] as AppInstance[]
    this.games = [] as GameInstance[]

    switch (platform) {
      case 'windows':
        this.factory = new WindowsStore
        break
      case 'osx':
        this.factory = new OSXStore
        break
      default:
        throw new Error(`Not know this platform ${platform}`)
    }
  }

  private getInstance(type: string, name: string): any {
    let instance: any
    switch (type) {
      case 'game':
        instance = this.games
        break
      case 'app':
        instance = this.apps
        break
      default:
        throw new Error(`Not found type is ${type}`)
    }

    return instance.find(inst => inst.getInfo().name === name)
  }

  run(type: string, name: string): void {
    const target = this.getInstance(type, name)
    if (typeof target === 'undefined') {
      throw new Error(`${type} instance with name ${name} not found`)
    }

    const { targetType } = target.getInfo()
    switch (targetType) {
      case 'game':
        target.play()
        break
      case 'app':
        target.start()
        break
      default:
        throw new Error(`Not found type: ${targetType}`)
    }
  }

  joinGame(name: string): void {
    const newGame = this.factory.createGame(name)
    this.games.push(newGame)
  }

  buyApp(name: string, deposit: number): void {
    const newApp = this.factory.createApplication(name)
    if (newApp.buyApp(deposit)) this.apps.push(newApp)
  }

  printInfo(type: string, name: string): void {
    const target = this.getInstance(type, name)
    if (typeof target === 'undefined') {
      throw new Error(`${type} instance with name ${name} not found`)
    }

    target.printInformation()
  }
}